src/main/java/frc/robot/subsystems/SwerveSubsystem.java

```diff
@@ -0,0 +1,191 @@
+    }
+
+    public SwerveModuleState[] convertToModuleStates(double xTranslation, double yTranslation, double rotation) {
+        //Takes axis input from joysticks and returns an array of swerve module states
+
+        double x = yTranslation; //Intentional, x in swerve kinematics is y on the joystick
+        double y = xTranslation;
+        double r = rotation;
+
+        //Map to speeds in meters/radians per second
+        x *= DriveConstants.kPhysicalMaxSpeed;
+        y *= DriveConstants.kPhysicalMaxSpeed;
+        r *= DriveConstants.kPhysicalMaxAngularSpeed;
+
+        //Log speeds used to construct chassis speeds
+        translationXOutputLog.append(x);
+        translationYOutputLog.append(y);
+        rotationOutputLog.append(r);
+
+        //Construct Chassis Speeds
+        ChassisSpeeds chassisSpeeds;
+        if(controlOrientationIsFOD){
+            //Field Oriented Drive
+            //TODO: test on real hardware, adding to gyro is supposed to eliminate skew of movement when rotating
+            chassisSpeeds = ChassisSpeeds.fromFieldRelativeSpeeds(x, y, r, this.getRotation2d().plus(new Rotation2d((r * Robot.kDefaultPeriod)/2)));
+        } else {
+            //Robot Oriented Drive
+            chassisSpeeds = new ChassisSpeeds(x, y, r);
+        }
+        SmartDashboard.putString("chassis speeds",chassisSpeeds.toString());
+        //Convert Chassis Speeds to individual module states
+        return DriveConstants.kDriveKinematics.toSwerveModuleStates(chassisSpeeds);  
+    }
+
+    private void normalizeDrive(SwerveModuleState[] desiredStates, ChassisSpeeds speeds){
+        //Find magnitude of translation input, map to a scale of 1 in order to be comparable to rotation
+        double translationalK = Math.hypot(speeds.vxMetersPerSecond, speeds.vyMetersPerSecond) / DriveConstants.kPhysicalMaxSpeed;
+        //Find magnitude of rotation input, map to a scale of 1 in order to be comparable to translation
+        double rotationalK = Math.abs(speeds.omegaRadiansPerSecond) / DriveConstants.kPhysicalMaxAngularSpeed;
+        //Use the larger of the two magnitudes for scaling
+        double k = Math.max(translationalK, rotationalK);
+      
+        //Find the how fast the fastest spinning drive motor is spinning                                       
+        double realMaxSpeed = 0;
+        for (SwerveModuleState moduleState : desiredStates) {
+          realMaxSpeed = Math.max(realMaxSpeed, Math.abs(moduleState.speedMetersPerSecond));
+        }
+        
+        if(realMaxSpeed != 0){
+            //Map input magnitude back to speed in meters per second, divide by real speed to create scale
+            double scale = Math.min(k * DriveConstants.kPhysicalMaxSpeed / realMaxSpeed, 1);
+            //Desaturate speeds using that scale
+            for (SwerveModuleState moduleState : desiredStates) {
+              moduleState.speedMetersPerSecond *= scale;
+            }
+        }
+    }
```

Comment:
Any case without rotations will have a ```k``` that is equal to realMaxSpeed/kPhysicalMaxSpeed, which will cause ```scale``` to always be 1. 

Can you provide more clarity about what you want this method to do that SwerveDriveKinematics.desaturateWheelSpeeds doesn't already provide? Removing k from the scale equation should fix the scaling issue, but then it is equivalent to desaturateWheelsSpeeds.

Reply:
The method was shared by a team on Chief Delphi, saying that it helped with oversaturation of the wheels when rotational and translational input were provided at once
I should probably just revert to WPI desaturation.

Reply:
But WPI's method still seems to oversaturate the wheels when full translational input on x and y is applied

Reply:
Did you take it from [here](https://github.com/TripleHelixProgramming/RapidReact/blob/b82c882b6c0f098bee2e87e16669d3ad8207aef9/src/main/java/frc/robot/drive/Drivetrain.java#L277)?

It looks like its used in conjunction with desaturateWheelSpeeds, not replacing it.

Reply:
Yes, that's what I used
I'll change our code to use it in conjunction with WPI's desaturate